### PR TITLE
spirv-fuzz: Fuzzer passes to add local and global variables

### PR DESCRIPTION
**Pull request by afd**
Adds two new fuzzer passes to add variables to a module: one that adds
Private storage class global variables, another that adds Function
storage class local variables.

### CODE REVIEW

File: source/fuzz/fuzzer_pass.cpp [hunk 1]
```diff
@@ -254,5 +293,152 @@ uint32_t FuzzerPass::FindOrCreateGlobalUndef(uint32_t type_id) {
...
+            GetIRContext()->get_constant_mgr()->GetConstantFromInst(
+                GetIRContext()->get_def_use_mgr()->GetDef(field_constant_id)));
+      }
+      return FindOrCreateCompositeConstant(
+          *type_instruction, field_zero_constants, field_zero_ids);
+    }
+    case SpvOpTypeVector: {
+      return GetZeroConstantForHomogeneousComposite(
+          *type_instruction, type_instruction->GetSingleWordInOperand(0),
+          type_instruction->GetSingleWordInOperand(1));
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass.cpp]: Nit: this looks identical to `case SpvOpTypeMatrix:` above. Could combine?

File: source/fuzz/fuzzer_pass.cpp [hunk 2]
```diff
@@ -254,5 +293,152 @@ uint32_t FuzzerPass::FindOrCreateGlobalUndef(uint32_t type_id) {
...
+
+uint32_t FuzzerPass::GetZeroConstantForHomogeneousComposite(
+    const opt::Instruction& composite_type_instruction,
+    uint32_t component_type_id, uint32_t num_components) {
+  std::vector<const opt::analysis::Constant*> zero_constants;
+  std::vector<uint32_t> zero_ids;
+  uint32_t zero_component = FindOrCreateZeroConstant(component_type_id);
+  const opt::analysis::Constant* registered_zero_component =
+      GetIRContext()->get_constant_mgr()->GetConstantFromInst(
+          GetIRContext()->get_def_use_mgr()->GetDef(zero_component));
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass.cpp]: I don't quite get why GetDef is needed here, but maybe I would if I had the code checked out.

File: source/fuzz/fuzzer_pass_add_local_variables.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    while (GetFuzzerContext()->ChoosePercentage(
+        GetFuzzerContext()->GetChanceOfAddingLocalVariable())) {
+      uint32_t base_type_id =
+          base_type_ids[GetFuzzerContext()->RandomIndex(base_type_ids)];
+      uint32_t pointer_type_id;
+      std::vector<uint32_t>& available_pointers =
+          base_type_to_pointer.at(base_type_id);
+      if (available_pointers.empty()) {
+        pointer_type_id = GetFuzzerContext()->GetFreshId();
+        available_pointers.push_back(pointer_type_id);
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_add_local_variables.cpp]: Might want to add a comment that this modifies the base_type_to_pointer map. This ensure that future loop iterations know that this type exists; it is not useful for this iteration. Could also rename it to available_pointers_to_base_type.

File: source/fuzz/protobufs/spvtoolsfuzz.proto [hunk 4]
```diff
@@ -516,6 +517,31 @@ message TransformationAddGlobalVariable {
...
+  // The id of the function to which the local variable should be added
+  uint32 function_id = 3;
+
+  // Initial value of the variable
+  uint32 initializer_id = 4;
+
+  // True if and only if the value of the variable should be regarded, in
+  // general, as totally unknown and subject to change.  Stores to such
+  // variables can be performed in an arbitrary fashion.
+  bool value_is_arbitrary = 5;
```
Comment (paulthomson) [inline source/fuzz/protobufs/spvtoolsfuzz.proto]: Could clarify that this is not like OpUndef (I think). It is more like: the fuzzer may add writes to this variable at any point.

File: source/fuzz/fuzzer_pass.cpp (hunk 2 above)
Comment (paulthomson) [reply source/fuzz/fuzzer_pass.cpp]: Oh it is because you need an Instruction*, right?

File: source/fuzz/transformation_add_local_variable.cpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  switch (initializer_instruction->opcode()) {
+    case SpvOpConstant:
+    case SpvOpConstantComposite:
+    case SpvOpConstantFalse:
+    case SpvOpConstantNull:
+    case SpvOpConstantTrue:
+      break;
+    default:
+      return false;
+  }
```
Comment (paulthomson) [inline source/fuzz/transformation_add_local_variable.cpp]: spvOpcodeIsConstant exists and has more cases, but perhaps that is fine.

File: source/fuzz/fuzzer_pass.cpp (hunk 1 above)
Comment (afd) [reply source/fuzz/fuzzer_pass.cpp]: Done.

File: source/fuzz/protobufs/spvtoolsfuzz.proto (hunk 4 above)
Comment (afd) [reply source/fuzz/protobufs/spvtoolsfuzz.proto]: Done.

File: source/fuzz/transformation_add_local_variable.cpp (hunk 5 above)
Comment (afd) [reply source/fuzz/transformation_add_local_variable.cpp]: Done (there is also SpvOpcodeIsComposite, which I should probably use elsewhere in the code base ...).

File: source/fuzz/fuzzer_pass_add_local_variables.cpp (hunk 3 above)
Comment (afd) [reply source/fuzz/fuzzer_pass_add_local_variables.cpp]: Good idea - done (and similar in fuzzer_pass_add_global_variables.cpp)

**Pull request merged**